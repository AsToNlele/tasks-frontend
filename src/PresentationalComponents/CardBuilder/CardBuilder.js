import React from 'react';
import { Card, CardBody, CardFooter, CardTitle } from '@patternfly/react-core';
import propTypes from 'prop-types';

const renderContentType = (data, item, actionFuncs) => {
  let returnedContent;
  let actionFunc = actionFuncs?.find((func) => {
    if (item.actionFunc === Object.keys(func)[0]) {
      return func[item.actionFunc];
    }
  });

  typeof item.content === 'function'
    ? (returnedContent = item.content(
        item.match === 'all' ? data : data[item.match],
        actionFunc?.[item.actionFunc]
      ))
    : (returnedContent = data[item.match]);

  return returnedContent;
};

const renderContent = (data, content, actionFuncs) => {
  return content.map((item, index) => {
    return (
      <div key={`${item.match}-${index}`} className={item.classname}>
        {renderContentType(data, item, actionFuncs)}
      </div>
    );
  });
};

const CardBuilder = ({
  actionFuncs,
  data,
  cardClass,
  cardHeader,
  cardBody,
  cardFooter,
}) => {
  return (
    <Card className={cardClass}>
      <CardTitle className={cardHeader.classname}>
        {renderContent(data, cardHeader.contents)}
      </CardTitle>
      <CardBody className={cardBody.classname}>
        {renderContent(data, cardBody.contents)}
      </CardBody>
      <CardFooter className={cardFooter.classname}>
        {renderContent(data, cardFooter.contents, actionFuncs)}
      </CardFooter>
    </Card>
  );
};

CardBuilder.propTypes = {
  actionFuncs: propTypes.array,
  data: propTypes.object,
  cardClass: propTypes.string,
  cardHeader: propTypes.object,
  cardBody: propTypes.object,
  cardFooter: propTypes.object,
};

export default CardBuilder;
