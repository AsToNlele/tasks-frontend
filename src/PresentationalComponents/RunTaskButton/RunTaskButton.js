import React from 'react';
import { Button } from '@patternfly/react-core';
import propTypes from 'prop-types';

const RunTaskButton = ({
  classname,
  task,
  isFirst,
  openTaskModal,
  variant,
}) => {
  return (
    <Button
      aria-label={`${task.slug}-run-task-button`}
      className={classname}
      variant={variant}
      onClick={() => openTaskModal(task)}
    >
      {isFirst ? 'Run task' : 'Run task again'}
    </Button>
  );
};

RunTaskButton.propTypes = {
  classname: propTypes.string,
  task: propTypes.object,
  isFirst: propTypes.bool,
  openTaskModal: propTypes.func,
  variant: propTypes.string,
};

export default RunTaskButton;
