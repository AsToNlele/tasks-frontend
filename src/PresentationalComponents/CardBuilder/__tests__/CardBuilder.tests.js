import React from 'react';
import { render } from '@testing-library/react';
import { Flex, FlexItem } from '@patternfly/react-core';

import CardBuilder, { CardBuilderContent } from '../CardBuilder';
import RunTaskButton from '../../RunTaskButton/RunTaskButton';
import { availableTasksTableItems } from '../../../Utilities/hooks/useTableTools/Components/__tests__/TasksTable.fixtures';

describe('TasksTables', () => {
  let props;

  beforeEach(() => {
    props = {
      cardClass: '',
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      <CardBuilder {...props}>
        <CardBuilderContent content="Task title" type="title" />
        <CardBuilderContent content="Task description" type="body" />
        <CardBuilderContent
          content={
            <Flex direction={{ default: 'column' }}>
              <FlexItem>
                <a href="#">Download preview of playbook</a>
              </FlexItem>
              <FlexItem>
                <RunTaskButton
                  isFirst
                  openTaskModal={jest.fn()}
                  task={availableTasksTableItems[0]}
                  variant="primary"
                />
              </FlexItem>
            </Flex>
          }
          type="footer"
        />
      </CardBuilder>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
