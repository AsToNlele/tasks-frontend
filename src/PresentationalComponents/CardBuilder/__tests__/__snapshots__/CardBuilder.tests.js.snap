// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardBuilder should render correctly 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      Task title
    </div>
    <div
      class="pf-c-card__body"
    >
      Task description
    </div>
    <div
      class="pf-c-card__footer"
    >
      <div
        class="pf-l-flex pf-m-column"
      >
        <div
          class=""
        >
          <a
            href="#"
          >
            Download preview of playbook
          </a>
        </div>
        <div
          class=""
        >
          <button
            aria-disabled="false"
            aria-label="undefined-run-task-button"
            class="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            Run task
          </button>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
